src/murmur/Messages.cpp

```diff
@@ -155,6 +155,61 @@ bool isChannelEnterRestricted(Channel *c) {
 	return false;
 }
 
+/// Process the Capabilities message, used to negotiate voice protocol.
+/// See the doc in Mumble.proto for details about this message.
+///
+/// @param uSource ServerUser that receives this message.
+/// @param msg The message itself.
+void Server::msgCapabilities(ServerUser *uSource, MumbleProto::Capabilities &msg) {
+	MSG_SETUP(ServerUser::Connected);
+	int custom_index = 0;
+	for (int i = 0; i < msg.supported_protocols_size(); ++i){
+		auto protocol_type = msg.supported_protocols(i);
+		switch (protocol_type) {
+			case MumbleProto::Capabilities_VoiceProtocol_UDP_OCB2:
+				uSource->qlSupportedVoiceProtocols.append(VoiceProtocol(VoiceProtocolType::UDP_OCB2));
+				break;
+			case MumbleProto::Capabilities_VoiceProtocol_CUSTOM:
+				uSource->qlSupportedVoiceProtocols.append(VoiceProtocol(msg.custom_protocols(custom_index)));
+				++custom_index;
+				break;
+			default:
+				continue;
+		}
+	}
+	uSource->vptVoiceProtocolType = VoiceProtocolType::UNDEFINED;
+	for (int i = 0; i < Meta::mp.qlAllowedVoiceProtocolTypes.count()
+		&& uSource->vptVoiceProtocolType == VoiceProtocolType::UNDEFINED; ++i) {
+
+		for (int j = 0; j < uSource->qlSupportedVoiceProtocols.count(); ++j) {
+			if (uSource->qlSupportedVoiceProtocols[j].protocolType == Meta::mp.qlAllowedVoiceProtocolTypes[i]) {
+				uSource->vptVoiceProtocolType = Meta::mp.qlAllowedVoiceProtocolTypes[i];
+				break;
+			}
+		}
+	}
+
+	uSource->initializeCipher();
+
+	MumbleProto::Capabilities mpc;
+	switch (uSource->vptVoiceProtocolType) {
+		case VoiceProtocolType::UDP_OCB2:
+			mpc.add_supported_protocols(MumbleProto::Capabilities_VoiceProtocol_UDP_OCB2);
+			break;
+		default:
+			mpc.add_supported_protocols(MumbleProto::Capabilities_VoiceProtocol_UNDEFINED);
+	}
+	sendMessage(uSource, mpc);
+
+	if (uSource->sState == ServerUser::Authenticated) {
+		MumbleProto::CryptSetup mpcrypt;
+		mpcrypt.set_key(uSource->csCrypt->getRawKey());
+		mpcrypt.set_server_nonce(uSource->csCrypt->getEncryptIV());
+		mpcrypt.set_client_nonce(uSource->csCrypt->getDecryptIV());
+		sendMessage(uSource, mpcrypt);
+	}
```

Comment:
Although the Capabilities message is sent before Authenticate, only send CryptState after the user is authenticated. 

Actually, when processing Authenticate message, a CryptState packet will be send in the middle, so we can just leave the CryptState there.

After the user is authenticated, nobody sends CryptState for me, so I have to send it here.